File: js/programmers/크레인 인형뽑기 게임.js
function solution(board, moves) {
  const stack = [];
  const len = board[0].length;
  let result = 0;
  for (let move of moves) {
    let i = 0;
    while (i < len) {
      if (board[i][move - 1] !== 0) {
        if (
          stack.length > 0 &&
          stack[stack.length - 1] === board[i][move - 1]
        ) {
          stack.pop();
          result += 2;
        } else {
          stack.push(board[i][move - 1]);
        }
        board[i][move - 1] = 0;
        break;
      }
      i++;
    }
  }

  return result;
}

console.log(
  solution(
    [
      [0, 0, 0, 0, 0],
      [0, 0, 1, 0, 3],
      [0, 2, 5, 0, 1],
      [4, 2, 4, 4, 2],
      [3, 5, 1, 3, 1],
    ],
    [1, 5, 3, 5, 1, 2, 1, 4],
  ),
);

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int n = board.length;
        int[] arr = new int[n];
        Arrays.fill(arr, -1);
        
        for (int i=0; i<n; i++) {
            for (int j=0; j<n; j++) {
                if (board[j][i] != 0) {
                    arr[i] = j;
                    break;
                }
            }
        }
        
        
        Stack<Integer> st = new Stack<>();
        int result = 0;
        
        for (int move : moves) {
            if (arr[move-1] == -1) {
                continue;
            }
            
            int v = board[arr[move-1]][move-1];
            if (arr[move-1] == n-1) {
                arr[move-1] = -1;
            } else {
                arr[move-1] += 1;
            }
            
            if (!st.isEmpty() && v == st.peek()) {
                st.pop();
                result += 2;        
            } else {
                st.push(v);
            }
        }
        
        return result;
    }
}